parking/shared/ws_models.py

```diff
@@ -0,0 +1,91 @@
+import json
+from enum import IntEnum
+import attr
+from parking.shared.location import Location
+from parking.shared.util import ensure, validate_non_neg
+from parking.shared.rest_models import ParkingLot
+
+
+class WebSocketMessageType(IntEnum):
+    LOCATION_UPDATE = 1
+    PARKING_REQUEST = 2
+    PARKING_ALLOCATION = 3
+    PARKING_ACCEPTANCE = 4
+    PARKING_REJECTION = 5
+    PARKING_DEALLOC = 6
+    PARKING_CANCEL = 7
+
+
+def deserialize_ws_message(data: str):
+    json_data = json.loads(data)
+    if '_type' not in json_data:
+        raise ValueError('Missing _type')
+
+    _type = json_data.pop('_type')
+
+    if _type not in [e.value for e in WebSocketMessageType]:
+        raise ValueError('Invalid _type: {}'.format(_type))
+
+    if _type == WebSocketMessageType.LOCATION_UPDATE:
```

Comment:
From a code quality perspective this could perhaps be a dictionary outside the function like `message_types = {LOCATION_UPDATE: LocationUpdateMessage, ...}` then do `return message_types[_type](**json_data)` in the code.

Some may prefer the if's though.

Reply:
I was thinking about doing it that way, but prefer not having to do a dictionary lookup on each WebSocket message

Reply:
Dictionary lookups are faster than chained if-statements :)

Reply:
I'll switch over then, I thought that the cache unfriendliness of the dict would make it slower than the if, but I guess the ifs will take some computation time...

Either way, I agree that it looks nicer so will switch over.